### ISSUE TITLE
Standard(s) for in-process propagation

### ISSUE THREAD

**Issue by yurishkuro** [opentracing/specification#23]
OpenTracing 1.x leaves the question of in-process context propagation entirely in the hands of the applications themselves, which substantially reduces the usefulness of the standard since different frameworks instrumented with OT do not have a way to exchange context.

Standards for in-process context propagation inevitably have to be specific to individual languages. This is more of a placeholder issue. It is also somewhat similar to #22 (possible part of it), but can be done independently.

**Comment by wu-sheng** [comment]
In-process propagation, and support different frameworks instrumented, seems very hard, as I known.
And different platforms have different characteristics.

In Java, same Context in ThreadLocal, and provide inter-thread propagation spec, maybe a effective way.

Keep me in the loop. :)

**Comment by adriancole** [comment]
Last july @Xylus and @eirslett did a couple day spike on context.
Assets from those meetings are here;
https://drive.google.com/drive/u/1/folders/0B0tSnQT3uGdAfndwS3RKVnMtM0ZKMWU4R1dicTNEXy1NSlBJaTlJQ05abGVneTlucW5HdFk

Eirik ended up spiking an effort towards that later,
https://github.com/DistributedTracing/continuation-local-storage-jvm/pull/1

Ironically, last december, iirc, distributed context propagation was
the first name of the OpenTracing group, so not too surprising it is
here.
Since then, I think uber had some work towards it naming the group
https://github.com/openctx

More recently, recently google are changing their own context to
decouple storage from it. https://github.com/grpc/grpc-java/pull/2461
Also, other tools like Log4J2 have added hooks to interop with other's
context storage engines.

As mentioned in another issue, there's some layering concern, for
example if tracing is the right place to define context (as they are
often multi-purpose and below tracing).
That said, whatever hooks there are tracing is at least a primary
consumer, and interesting to see a tracking issue arise.

**Comment by yurishkuro** [comment]
@adriancole what I find peculiar is that several similar concepts of context storage/provider I've seen (including Jaeger's) _pretend_ that they are just an interface, but in fact they cannot work with anything but a thread-local based implementation. For example, TChannel (which is a form of RPC framework) can be configured with a [TracingContext](https://github.com/uber/tchannel-java/blob/master/tchannel-core/src/main/java/com/uber/tchannel/tracing/TracingContext.java), as a way of abstracting the actual mechanism of in-process propagation, but the mere fact that a context object is not passed to RPC methods but instead extracted via that abstraction essentially means the abstraction must be implemented via thread-locals.

NB: the TracingContext in Jaeger and TChannel is to a Span what gRPC Storage is to a Context. I.e. the same separation of _how_ something is propagated and _what_ is being propagated. They are just smaller in scope and only propagate a single Span.

**Comment by adriancole** [comment]
> @adriancole what I find peculiar is that several similar concepts of context storage/provider I've seen (including Jaeger's) pretend that they are just an interface, but in fact they cannot work with anything but a thread-local based implementation.

interesting point for JVM tech. yeah different types of thread locals,
wrappers of thread locals, or thread-local references to something
else.. all have thread locals in there somewhere. Maybe there's some
subtle (or native) option I'm also unaware of. @raphw @tmontgomery you
know magic ( or at least where bodies are often buried ) .. do you
know something available on JVM that can be used to store incidental
trace state which isn't directly or indirectly a thread local?

**Comment by raphw** [comment]
As a matter of fact, I have written a small [thread-local-ish utility](https://github.com/raphw/weak-lock-free) for this exact purpose. The problem with thread locals is that it only allows access from within a `Thread`. It is not possible to set an object from without a `Thread` even if the latter is known.

Also, I have started migrating away from the above `DetachedThreadLocal` and rather use the `WeakConcurrentMap`. The latter allows to define a form of key object that represents a state. This state is automatically garbage collected once the key is eligible for collection. Propagation works by sharing such keys where the context is removed once a context has become obsolete.

All in all, the problem always boils down to managing a form of global-state what is always tedious.

**Comment by wu-sheng** [comment]
@adriancole  what is incidental trace state? Why should be stored in something like threadlocal?

**Comment by wu-sheng** [comment]
To all, threadlocal can't solve all problems, somthing like async process model will bring us more.

For example, [Disruptor](https://github.com/LMAX-Exchange/disruptor) is a very useful in-memory async lib. If application use this, all context base on threadlocal will fail without doubt. 

This is very important for my exp.

**Comment by adriancole** [comment]
>
> @adriancole <https://github.com/adriancole> what is incidental trace
> state? Why should be stored in something like threadlocal?
>
sorry poor choice of words. Most users aren't aware of trace state as it is
implicitly passed around on their behalf (of course some are). The usual
implicit mechanism in Java is a thread local, though it is problematic in
async libraries. For example, usually you have to manually re-attach the
state when some accident of thread scheduling loses it. Ex the below from
Brave + RxJava

public Action0 onSchedule(final Action0 action) {
final ServerSpanThreadBinder binder = brave.serverSpanThreadBinder();
final ServerSpan span = binder.getCurrentServerSpan();
return new Action0() {
@Override
public void call() {
binder.setCurrentSpan(span);
action.call();
}
};
}
https://github.com/smoketurner/dropwizard-zipkin/blob/master/zipkin-rx/src/main/java/com/smoketurner/dropwizard/zipkin/rx/BraveRxJavaSchedulersHook.java


One of the reasons I asked about alternatives to thread locals was to
figure out what current practices exist, as it is helpful to scope what
might end up in a spec or library here or elsewhere.

make sense?

**Comment by wu-sheng** [comment]
@adriancole, yes, that is my concern.

**Comment by wu-sheng** [comment]
But I think only change threadlocal to something similar is not explicit enough to someone, who is not a tracer developer.

Provide some APIs, like carrier, inject, extract for rpc, are better choice for me. They will be easy to understand.

**Comment by cwe1ss** [comment]
In terms of the specification, the more interesting question for me is the contract of this API:

* Should it be possible to store just one span in the context?
  * This would probably be e.g. the RPC server span and all lower-level components would be child spans of this one span without any nesting.
* Should it be implemented as a stack?
  * How do you know on POP that you got "your" span? (in case a child component failed to close/pop its span or called pop too often)
  * How does a component know that it is the outermost component and that it should close any open spans. Example: The RPC server creates a "root span" and adds it to the stack, a lower-level component (e.g. a DB access library) opens a child span but the code failed with an exception. The request exception handler now needs to know that the DB span must be closed.
* Should it be implemented as a list or a tree to allow for more reference types?

It might be easier to decide which language-specific storage construct might be the right one, when we know which features we want it to support.

**Comment by sjoerdtalsma** [comment]
I am creating a library called 'context-propagation' meant purely to propagate any (threadlocal) context implementation from a parent-thread to any background thread in a stack-like fashion, using the Autocloseable try-with-resources mechanism to allow verified stack unrolling (and automatic skipping if a pop is missed).
It revolves around a Context and ContextManager interface (each with only two methods) and a utility to create a snapshot from all known ContextManager implementations at once.
If you're interested, please take a look here and let me know what you think:
https://github.com/talsma-ict/context-propagation

(we needed it for security propagation, but are now looking into adding opentracing and would like to re-use the mechanism that is currently in-place that serves us well).

**Comment by yurishkuro** [comment]
@cwe1ss I think the only requirement is "I need to get the current span". The fact that Java implementation often uses stack is a side-effect of thread-local implementation, since you're using the same API to retrieve different spans at different times, and stack is the only thing that makes sense for that.

**Comment by cwe1ss** [comment]
@yurishkuro there's also the logical requirement to "set the current span". Do you think this is implicit - meaning that every span that is created is automatically "the current span" or do you see this as an explicit feature/step? 

In other words, do you think that the context should be part of the `tracer` interface in which a call to `StartSpan` would automatically store the span in the context and something new like `tracer.CurrentSpan` would return it from the context? Or would you prefer a new interface like `traceContext` that must be called explicitly by the user after a span has been created - e.g. `traceContext.SetCurrentSpan` and `traceContext.GetCurrentSpan`?

The former would make for a nice user-friendly API and the latter seems to be closer to what we have now in jaeger-context and my C# contrib library. But both have advantages and disadvantages that need further discussions of course.

**Comment by yurishkuro** [comment]
Sorry, you're right, it needs get and set fort current span.

**Comment by bensigelman** [comment]
@sjoerdtalsma thanks for sharing the context propagation lib... nice to see something that's already in use and not pre-coupled to a distributed tracing system per se.

@yurishkuro @cwe1ss I am prototyping a few approaches to make this pluggable... I would like OT to support in-process context managers without getting into the business of doing all of the work (since there are so many approaches and each has its advantages/disadvantages).

@cwe1ss as far as the stack concept is concerned: in Dapper we basically had each in-process Context object keep a reference to its parent Context; when `finish()` was called (or its moral equivalent... the naming was different), the Context would re-install its parent Context in the thread-local storage. So there was essentially a linked list of Context objects within the process.

**Comment by sjoerdtalsma** [comment]
@bensigelman For what it's worth, I heavily depend on the the java.util.ServiceLoader to make my library pluggable.
The abstract threadlocal implementation has the same 'linked stack' structure, except it tries to correct out-of-order closing as well.

Details on the out-of-order closing: If the closed context isn't 'current', current isn't touched.
While the parent is already-closed, it is 'skipped' in the search for the restored 'current'.

**Caution:** I pushed a major package refactoring yesterday (on _develop_) in preparation of 1.0 of the library, but the github website does not yet reflect this! I suspect clone or pull probably will.

**Comment by wu-sheng** [comment]
@sjoerdtalsma , can you post the site of **java-globaltracer** project. I have suggest earlier in https://github.com/opentracing/opentracing-java/issues/63. ServiceLoader to get a tracer maybe a good way. Even though, some tracer need arguments to init, like Jaeger Tracer by @yurishkuro told, they are not suitable for the module.

But I like this type of java tracer factory/manager very much.

**Comment by sjoerdtalsma** [comment]
@wu-sheng @adriancole , I pushed an initial attempt (with a working unit test using the MockTracer) to github yesterday evening:
https://github.com/talsma-ict/java-globaltracer

I would love to donate it to the 'contrib' git repository and have it published in Maven central under 'io.opentracing.contrib' but only after a thorough review process. Since I'm relatively new to the opentracing community I have no experience on the best way to go forward from here...

**Comment by wu-sheng** [comment]
@sjoerdtalsma @adriancole  , maybe we should create a new issue to discuss this? 

I review the project generally,  let's leave the detail discuss later. 

But at first, jdk level should be fixed. Using `Optional<T>` led to needing jdk 1.8+, but many applications based on jdk 1.6+. I think you should implement your design on jdk 1.6, maybe the code is less grace, but it will easier to use.

**Comment by adriancole** [comment]
>
> @sjoerdtalsma <https://github.com/sjoerdtalsma> @adriancole
> <https://github.com/adriancole> , maybe we should create a new issue to
> discuss this?
>
One of the reasons we discussed thread local was yuri's question of whether
we can assume thread locals are used or not. as far as I can tell, we
cannot make that assumption. If that steers the api in a particular way,
then the examples were fruitful.

I agree that it is probably helpful to keep weeds or investigations on
different threads as doing so here might be overwhelming. probably best in
the actual repo of the code in question?

**Comment by sjoerdtalsma** [comment]
@wu-sheng @adriancole I welcome feedback on the design and implementation choices by github issues.
https://github.com/talsma-ict/java-globaltracer
I'm also willing to transfer this repository to https://github.com/opentracing-contrib if you think this contribution is worthwhile.

**Comment by wu-sheng** [comment]
@sjoerdtalsma , transfer to https://github.com/opentracing-contrib, LGTM. @yurishkuro  @adriancole @bensigelman , what's your opinion? 

Maybe accept it, discuss on the repo, and let's see what is going on later?

**Comment by bensigelman** [comment]
@sjoerdtalsma my preference would be to

1. create (empty, ASF2.0) repos on opentracing-contrib for the pieces you'd like to move over
1. do a PR for the initial contents where we discuss the finer points of the code itself (and I would def rather *not* try to do that all here in this thread!)
1. merge those PRs as soon as they're ready, etc.

**Comment by wu-sheng** [comment]
@bensigelman , can you add me to the https://github.com/opentracing-contrib org. I am interested in @sjoerdtalsma's project.

**Comment by wu-sheng** [comment]
@sjoerdtalsma , let's me know, after you create a repo, and do a PR

**Comment by sjoerdtalsma** [comment]
@bensigelman I agree on all three of your points.
re. 1: I cannot create this repo in opentracing-contrib, so it would be great is anyone with the power to do so could create a new ASF2.0 repo called 'java-globaltracer' or something similar. I'll take care of step 2 from there.

**Comment by wu-sheng** [comment]
@bensigelman , maybe you should create a repo for @sjoerdtalsma

**Comment by bensigelman** [comment]
@sjoerdtalsma: https://github.com/opentracing-contrib/java-globaltracer/invitations gives you admin permissions... maybe we can start with a PR of the existing code?

**Comment by bensigelman** [comment]
(See also: https://github.com/opentracing-contrib/java-globaltracer/pull/1)

**Comment by mabn** [comment]
> @cwe1ss I think the only requirement is "I need to get the current span"

> @yurishkuro there's also the logical requirement to "set the current span"

One thing I don't want is having to create a second instrumentation for non-tracing context propagation. 
For example - imagine an http service which authenticates a user. It obtains a *User* instance. It's natural that to want to access this instace from within the request processing "flow" and so it has to propagate (implicitly or explicitly) across various calls, just like tracing context.

Another good example is having a per-request cache.

Right now opentracing api does not support this - tags do not propagate and baggage values are strings (and propagate to other services).

So: **I'd like a way to to store objects in some process-scoped baggage / request-context**

Second thing that I believe is important is **playing nicely with MDC** from logging frameworks. Explicit support is not needed, but it should be convenient.

I know that one could say that tracer implementations should handle MDC integration and "request context" could be just a Map with span-id as the key, but it should be considered.

**Comment by JonathanMace** [comment]
Hi all,

Just dropping in here to post a message that I also posted to the Google Group.  I've recently been working on "Baggage Buffers" with Rodrigo at Brown.  Baggage Buffers, and more broadly the Tracing Plane, is the culmination of our ideas related to distributed tracing, metadata propagation, and Baggage.

I've just finished the end-to-end implementation and I'm working on testing, tutorials, examples, and documentation.  However, I thought I'd just jump the gun and let you all know about this now, because some of the concepts are pretty cool (baggage buffers itself, as well as the atom layer and baggage protocol).

https://github.com/JonathanMace/tracingplane

An actual release is going to follow in February, along with a whitepaper and research paper (plus much more extensive documentation and examples).

============


On a separate note, I'm annoyed that I missed this conversation and I will read through it and provide some comments.  Baggage Buffers supports process-scoped baggage with 'transient' datatypes and more generally, scoping for when variables are propagated and not propagated.

**Comment by wu-sheng** [comment]
@JonathanMace , seem several interesting things. But a little hard for me to understand all (My English may be not good enough). Still,  look forward your `Getting started`, `Simple example` and `Tutorial`. 

Thank you for your sharing.

**Comment by bensigelman** [comment]
I'd like to move this conversation along... I think this issue is the most pressing technical concern for OpenTracing as a project right now and I'm hoping to move the discussion forward a bit in the coming week or two.

So, since the last activity in this thread, @sjoerdtalsma has merged in his https://github.com/opentracing-contrib/java-spanmanager work. Per the README,

> Warning: This library is still a work in progress!

That said, he put a bunch of thought into it, and it makes an interesting reference point for the discussion here. Separately, I've been thinking about how we addressed this problem at Google and remembering what I liked and did not like about it.

TL;DR, for OpenTracing to achieve its goals, it's nearly a requirement for a programmer to start a new Span without a local var that identifies its parent. Otherwise there is this annoying handoff problem between API users and API implementors if said API is going to support OpenTracing. (There are approaches like https://github.com/opentracing-contrib/java-jdbi/blob/master/src/main/java/io/opentracing/contrib/jdbi/OpenTracingCollector.java#L185, but they're heavyhanded and still don't really solve the problem except for trivial situations where thread-local storage is sufficient)

Now, in order to satisfy the above (near-)requirement, OpenTracing needs to provide:
- access to an appropriate `Tracer`
- some way of accessing a parent SpanContext to reference in any newly-started Span

I would note that many existing solutions in this arena provide access to the active *Span*, not just the SpanContext... IMO that's more than is necessary and also creates complex corner-case lifetime issues around `Span.finish()` (remember that SpanContext objects are by-definition good-for-life and don't expire on `Span.finish()`).

Now, we could make an "official" OpenTracing Futures library, an "official" OpenTracing executor service, etc, etc... and maybe we will someday. But I would love to keep those things out of the core API and make them opt-in. That suggests some way to make the active-SpanContext-management code pluggable somehow.

Remembering that we're in the `specification` repo rather than the `java` repo, I don't want to overfit to Java... but it is helpful to make this concrete. So here's a rough sketch of a suggestion here:
- Introduce something like an `ActiveSpanContextProvider` interface. It would have one method, `get()`, which would return the active `SpanContext`, or `null` if none is known
- Introduce a `SpanLifetimeObserver` interface. It would have callback methods to inform about new Span start and finish events, but might only find out about the SpanContext
- `Tracer` gets an `addActiveSpanContextProvider` method, which requires an `ActiveSpanContextProvider` *and* a `SpanLifetimeObserver` implementation
- things like `startSpan` would have some sort of behavior where they would look through the SpanContext providers for a possible parent (not sure how to handle multiple parents if at all?)

Now, a Futures framework or executor wrapper can bind to any OpenTracing impl via these two interfaces. Users would have to make that binding at init-time, but then in their application code they could basically just write something like `globalTracer.buildSpan("foo").start()` and OpenTracing could fall back on the `ActiveSpanContextProvider` to figure out the parent/child relationship.

Thoughts?

PS: If the above is hard to understand I can write a gist or something. Mainly just doing a braindump at this point.

- `Tracer`

**Comment by wu-sheng** [comment]
First, on most sides, I support you about this. `ActiveSpanContextProvider` and `SpanLifetimeObserver` are good interface. 👍  +1 on adding these to spec 2.0

And I have to say, these interfaces are very google-naming-style. 😜

We can have more discussion about this feature.

First of all, `Tracer` have methods to add `ActiveSpanContextProvider` and `SpanLifetimeObserver`. And SpanBuilder has `asChildOf(SpanContext spanContext)`, I am not sure you want to remove this method, because you mention 

>But I would love to keep those things out of the core API and make them opt-in. That suggests some way to make the active-SpanContext-management code pluggable somehow

We will face two types of `context`, if the context-provider mechanism is active.  

If we do like activespan-repo did, such as`globalTracer.buildSpan("foo").start()` , we will need a set of new APIs. `NewTracer`'s methods are `static`. This is not like a plugin, but like an extend-API.

At this point, providing these two interfaces just as plugin, seems have a little problem. Hopefully, you already have some solutions on these.

**Comment by mabn** [comment]
1
`SpanLifetimeObserver` should also have callbacks for activation and deactivation of a span. It's required for example to fill thread-locals when a new threads starts execution in the context of existing span and for subsequent cleanup.

2
I'd still want support for "request context" (like baggage, but does not propagate outside of a single process and values can be arbitrary objects). I made some attempts to implement it and at least following is required:
- lifecycle callbacks for span start and activation
- at least "start" callback must have access to the parent span (or parent span context)
Then it can be achieved also with a separate "manager" which would track the hierarchy of spans and keep a mapping between spans and "request contexts"

But it would be great to add "request context" interface to the API.

3 
Java-specific: I played a bit with MDC integration. The tricky scenario is:
P and C denote parent and child threads, p and c parent and child span
```
P: activate(p)
P: MDC.put("x", 1)
P: spawn C
P: MDC.put("y", 2)
C: c = child_span_of(p)
C: activate(c)
C: assert MDC.get("x") == 1
C: assert MDC.get("y") == 2
```

Here both the start and activation happen in the Child thread and it's too late - there's no way to access MDC from the Parent thread. I'm not sure how to solve it without replacing underlying MDC implementation (MDCAdapter in case of Slf4j).
It's easy if child span starts in Parent thread (it's possible to create a copy of the current MDC and later set it in Child thread)

**Comment by bensigelman** [comment]
@mabn a few thoughts:

> SpanLifetimeObserver should also have callbacks for activation and deactivation of a span

I agree that such an observer needs to know about those events, but I *don't* think the Tracer would be the one delivering them... e.g., the Executor wrapper itself would intrinsically know about activation and deactivation.

> ... it would be great to add "request context" interface to the API.

This one is hard for me. On the one hand, I agree with you about its elegance and so on. On the other hand, OpenTracing gets dragged over the coals for solving (or trying to solve) too many problems at once. (E.g., @adriancole's refrain about whether logging needs to be so general-purpose, or [this issue](https://github.com/opentracing/opentracing-go/issues/127) from OT-Go) Then you have things like @JonathanMace's https://github.com/JonathanMace/tracingplane , which is kind of like what you're describing and much more.

My personal preference would be to remain *aware* of this sort of functionality as we design APIs and attempt to be roughly forwards-compatible (if not from a strict compilation standpoint, at least conceptually / holistically).

> ... MDC integration.

Yeah, MDC is one of those almost-but-not-quite things for me. I also don't know how to make it work in the general case without reimplementing.

**Comment by sjoerdtalsma** [comment]
I think the API should limit itself to Tracing and not try to include 'foreign' features such as RequestContext or MDC etc. That creates a huge (and expanding) scope for this.
Ideally, the API should easily integrate with other context propagation mechanisms.

Purely as example, I added Span propagation to my existing library, unfortunately I didn't have time for documentation or examples yet, but if you're open to looking at code here's an example with propagation of `Context[ServletRequest]` and `Context[Span]`, the latter delegating to any SpanManager it can find on the classpath.
- https://github.com/talsma-ict/context-propagation/tree/develop/context-propagation-opentracing-span
- https://github.com/talsma-ict/context-propagation/tree/develop/context-propagation-servletrequest

**Comment by sjoerdtalsma** [comment]
>Then you have things like @JonathanMace's https://github.com/JonathanMace/tracingplane , which is kind of like what you're describing and much more.

Very interesting concept! I'd love to integrate with something as common as that. However, maybe OpenTracing should merely provide the API's that _allow_ this integration and leave the integration up to the chosen solution?

**Comment by bensigelman** [comment]
@sjoerdtalsma

> I think the API should limit itself to Tracing and not try to include 'foreign' features such as RequestContext or MDC etc. That creates a huge (and expanding) scope for this.
Ideally, the API should easily integrate with other context propagation mechanisms.

That's essentially what I'm trying to get at here. OpenTracing should try to stay out of the business of "picking winners" for context propagation mechanisms. There are too many, and they have too many users. *However*, it would make instrumentation a lot easier to write if that binding could be decoupled from the actual line-by-line instrumentation code... i.e., if the binding could happen at main() or ServiceLoader init time. The proposed provider interface is a very rough attempt to get at something like this.

This week I hope to spell out a complete working example of how the binding would actually look in Java... maybe one for MDC, one for vanilla thread-local storage, and one for an executer wrapper?

**Comment by pauldraper** [comment]
> maybe one for MDC, one for vanilla thread-local storage, and one for an executer wrapper

An executor wrapper would be necessary for both MDC and ThreadLocal.

If you do MDC, realize there are two flavors: inherited (Log4j and old Logback) and non-inherited (new Logback). The later is the most sensical.

I'm very happy with java-spanmanager. My only recommendation would be to be able to set the global `SpanManager`, like with java-globaltracer.

**Comment by sjoerdtalsma** [comment]
@pauldraper 
>I'm very happy with java-spanmanager. My only recommendation would be to be able to set the global SpanManager, like with java-globaltracer.

Glad it's useful to you! I have something set up in my own repo like that: [GlobalSpanManager](https://github.com/talsma-ict/context-propagation/blob/master/opentracing-span-propagation/src/main/java/nl/talsmasoftware/context/opentracing/GlobalSpanManager.java).
Feel free to create an issue or pull request on java-spanmanager to discuss this in more detail.

**Comment by tedsuo** [comment]
Hi y’all! I’ve been thinking about this issue with @bhs. I think it might be helpful to break this problem down into two pieces – User APIs and Implementor APIs – and to sum up some conversations that have been happening in various threads. My main point is to try and draw a line between APIs that are used by applications and libraries (call them **User APIs** for now), vs APIs that are used by the writers of tracers and context managers (call them **Implementor APIs**). Hopefully this is a helpful framing of the context management problem.

## APIs for Implementors

For an API to be an Implementor API, it must only be called during tracer initialization, or internally between context managers and tracers. The tracer should not expose these APIs in any way. 

Things that happen **only** at init or within the internals of a Tracer have some nice properties:

- Can be part of `opentracing-java`, but are less critical to get right as they can be changed later without world-destroying effects. These API changes would only affect calls at application init and and not the zillions of instrumentation call sites within applications and libraries. Any code for in `opentracing-java` for maintaining the relationship between tracers and context managers, context management fallback, etc, can be improved while only affecting tracer and context manager authors, not application/library developers. 

- Can instead be specific to the tracer implementation, since you still have a handle to the tracer implementation at that point. Multiple tracers can consolidate on APIs for context management – because why reinvent `SpanPropagatingExecutorService` N times – but have room to reinvent the wheel if they have some fancier or experimental way of doing things.

We still have to solve these problems for implementors, but given how wide the topic of context management is, and that it may continue to grow with the future of Java, there is a lot to be said around keeping these APIs inaccessible from the outside, past Tracer initialization.

## APIs for Users

For normal instrumentation, you can wrap all this up and allow the user to call the existing `tracer.buildSpan().start()` off of a singleton-like tracer object. Indeed, that’s that goal!

The only context management that needs to be exposed via the Tracer API is whatever is needed to allow the user to do surgery around the edge cases, such @adriancole’s (example)[https://github.com/opentracing/specification/issues/23#issuecomment-264871406] of manually reactivating the span when RxJava loses track of it. So, potentially this new public Tracer API is smaller than the one the Tracer may need internally to interact with context managers. 

Given these constraints, is the following additional API on the tracer sufficient for this kind of user land surgery? This appears to cover all the use cases I have seen, but I may be missing something important. I’m calling this `ActiveSpanProvider` only because I had to pick a name. Also, I’m going with Span, not SpanContext, as I believe that debate has been settled.

```java
public interface ActiveSpanProvider {
  Span activeSpan();

  // caller must replace their span reference with the returned span
  Span activateSpan(Span span); 

  void deactivateSpan(Span span);
}

public interface Tracer extends ActiveSpanProvider {
// … existing API … //
}
```

Again, hopefully this framing was helpful. Also, I’d love feedback, especially examples of this “User API” being insufficient.

**Comment by wu-sheng** [comment]
@tedsuo @bensigelman ,  First, I love the `User APIs and Implementor APIs` point. This will be very helpful.  💯 

I assume, do you want to remove the `SpanContext` from **User APIs`? So, users can use `tracer.buildSpan().start()` to start a new span directly, and do not need care about `SpanContext`.

And I have to say, I have a little concern about `ActiveSpanProvider`. Maybe later, you can explain more about `activateSpan` and `deactivateSpan`, and give me some demos.

Another thing is, after we separated these two pieces, what is `opentracing-java` responsibility? `opentracing-java` is for User APIs, and `opentracing-tracer-java` is for Implementors APIs?

**Comment by mabn** [comment]
@tedsuo 

I think User API is insufficient as it does not allow to instrument Slf4j's MDC. The method names are fine, but the fact that they deal with "Span" is too limiting.

For MDC the instrumentation of a Runnable could look like this:
```
  static class InstrumentedRunnable implements Runnable {
    private Something context;
    private ActiveSpanProvider provider; 
    private Runnable wrappedRunnable;

    public InstrumentedRunnable(ActiveSpanProvider provider, Runnable wrappedRunnable) {
      context = provider.beforeActivation();
      this.wrappedRunnable = wrappedRunnable;
    }

    @Override
    public void run() {
      provider.activateSpan(context);
      try {
        wrappedRunnable.run();
      } finally {
        provider.deactivateSpan(context);
      }
    }
  }
```

The important part is that there are 3 places where libraries need to interact with the tracer and instrumentation. 
Let's consider using above InstrumentedRunnable to execute something in another thread. MDC instrumentation would be interested in:
1. capturing current state in original the thread/fiber/actor - `beforeActivation` in above example. MDC instrumentation would run `MDC.getCopyOfContextMap()` at this point
2. setting this state in the new thread/fiber/action - that's what `activateSpan` does. In case of MDC it is `MDC.setContextMap(capturedState)`. In case of a span it could be `provider.setActiveSpan(...)`
3. cleanup - `deactivateSpan`

As you can the methods required to instrument MDC are pretty similar, but they cannot deal with Spans but rather some object which contains active Span and additional "contexts" from all hooks/plugins which are interested in participating in the activation flow.

This is the case not just for MDC but in general for any functionality which maintains state related to a subset of spans - e.g. a request context, or some thread-local state.

**Comment by tedsuo** [comment]
Thank you for the example @mabn, it’s very helpful. Using a global Tracer with the ActiveSpanProvider API I proposed, your example could be written as follows:

```java
  static class InstrumentedRunnable implements Runnable {
    private Span span;
    private Runnable wrappedRunnable;

    public InstrumentedRunnable(Runnable wrappedRunnable) {
      this.span = GlobalTracer.get().activeSpan();
      this.wrappedRunnable = wrappedRunnable;
    }

    public InstrumentedRunnable(Span span, Runnable wrappedRunnable) {
      this.span = span;
      this.wrappedRunnable = wrappedRunnable;
    }

    @Override
    public void run() {
      tracer = GlobalTracer.get()
      span = tracer.activateSpan(span);
      try {
        wrappedRunnable.run();
      } finally {
	 tracer.deactivateSpan(span);
      }
    }
  }
```

Internally, the tracer has associated the span with the contents of `MDC.getCopyOfContextMap` when the span was created. On `Tracer.activateSpan` the Tracer calls `MDC.setContextMap` with this context. So the machinery for managing the relation between the Tracer and MDC/SomeContextManagerAPI is not exposed to the framework developer. So, the tracer developer manages Context, the framework developer manages Activate/Deactivate, and the application developer manages Start/Finish.

Thoughts?

**Comment by tedsuo** [comment]
Hi @wu-sheng , glad you like the User/Implementor distinction. Just to be clear, my comments above are only related to context propagation, I’m not proposing a huge change to the existing API.

- The current tracer can be accessed from the environment (AKA a `globalTracer`).
- The tracer includes an ActiveSpanProvider interface.
- Internally, the tracer interacts with context managers. 
- Activate/Deactivate is pulled out from Started/Finished. Start and Finish are states on a Span, Active/Inactive describes the relation between a span and the current context.
- There was some debate about whether context managers and tracers should use Spans or SpanContexts when they interact with each other. SpanContexts looked attractive because they are immutable, but their current API is too restrictive, and we don’t want to extend it. So we should use Spans instead. I'm not recommending we get rid of SpanContext.

My main point about User/Implementor is just some framing: we have Tracer developers, Framework developers, and Application developers, and their roles and APIs should ideally not bleed into each other. The Application developer interface is the most important to standardize, followed by the Framework developer, due to the number of call sites those APIs represents compared to internal Tracer APIs.

**Comment by wu-sheng** [comment]
I see your point. Distinction for Tracer developers, Framework developers, and Application developers, which are three important roles, when we build a trace eco-system.

The codes like these (from your demo), make easier to read/write (ignore span or context thing, right now). I hope OT can have improvements on that.
```java
@Override
    public void run() {
      tracer = GlobalTracer.get()
      span = tracer.activateSpan(span);
      try {
        wrappedRunnable.run();
      } finally {
	 tracer.deactivateSpan(span);
      }
    }
```

Only for my personal interest, I think only `Tracer developers` should focus on context; `Framework developers`, and `Application developers` only concern about context propagation when cross thread/process ( Maybe inject/extract mechanism ). This will make all things easier to use/understand.

**Comment by mabn** [comment]
@tedsuo I thought about it - hiding the "context" behind the Span interface - and it would work, but it would also cause unnecessary overhead. Here's why:

(I'll use MDC example)

1. To extract a snapshot of MDC context some action must be performed in the current thread (`MDC.getCopyOfContextMap()`) - it's not possible to change internals of MDC and have access to all contexts any time (it is possible with some reflection magic, but let's not go there).
2. The only place to do that with the proposed API is the `activeSpan()` method.
3. But I suspect that `activeSpan` will be called pretty often for reasons other than to jump threads. E.g. to add tags or logs to the active span. There's no way to indicate the purpose so every call will need to extract MDC context and wrap it into a new Span instance (it has to be a separate instance because thread safety).

On the other hand e.g. logback maintains MDC context as copy-on-write-map so taking a "snapshot" is basically free.

**Comment by wu-sheng** [comment]
> every call will need to extract MDC context and wrap it into a new Span instance

@mabn , why need wrap it into a new Span, every time? IMO, it just builds a new span, based on the given context( whatever who manage it ), this should be on the `Tracer APIs` head.

I think the point is not **hide** `context`, it is about whether need to know/use the `context` on the `User APIs` side.

**Comment by mabn** [comment]
You cannot store just one MDC context per span because of this:

```
Span span1 = provider.activeSpan()
MDC.put("A", "a")
Span span2 = provider.activeSpan()

// activate span1 in thread1
// activate span2 in thread2
```

Both threads will get MDC context with "A" because span1 and span2 are the same spans. But only thread2 should see newly added key "A". 
I wrote a sequential example, but the point is that it leads to race conditions. Another example would be:

```
MDC.put("client_id", 1)
Span parentSpan = provider.activeSpan()
// start new child thread which calls provider.activate(parentSpan);
MDC.put("client_id", 2);
provider.activeSpan()
```

In above example the client_id seen by child thread depends on on the order of execution - if child thread executed `activate` before second `provider.activeSpan()` it will see client_id == 1, otherwise 2.

Maybe there's some way to deal with without wrapping every time but I don't see it right now.

**Comment by sjoerdtalsma** [comment]
@mabn I don't understand your combination of MDC and OpenTracing api.
I think these are separate concerns that _may_ propagate together (or may not).
In my code, I have a specific Tracer delegate that updates some MDC fields when starting/finishing spans (e.g. span / trace id's etc).
This is combined with a specific SpanManager that does the same when Spans are managed/released.
In my opinion these are separate concerns which should not be mixed.

_by the way, is this becoming too Java specific? I doubt there's an MDC concept in all languages_

**Comment by mabn** [comment]
@sjoerdtalsma

I believe that it's a general concept of a _state_ with lifecycle bound to a particular concurrency model (threads in this case and thread-local state).

The point is that you need to create a copy of that _state_ before passing it to another thread/actor/channel/whatever.

Using `provider.activeSpan()` for both:
- extracting active span for activation in another thread/...
- and accessing current active span just to add tags/logs

forces the implementation to always create a copy of that _state_, even if the purpose is to just add tag to the active span. Maybe it's a good tradeoff for the simplicity of ActiveSpanProvider interface.

_this comment has been certified as "Java-free"_

**Comment by tedsuo** [comment]
@mabn, I see what you are saying. Users will be calling `provider.activeSpan` simply to grab the logger, and there may be a performance penalty if that method must do extra work to prep the span in some way for context propagation. 

However, I’d like to explore how real that is, because having two methods opens up other problems. Let’s say we have two methods, `provider.activeSpan` and `provider.beforeActivation`. If both methods return the active span, that opens the door to undefined behavior. AKA what happens when you call `activateSpan` on a span that has not been prepped by `provider.beforeActivation`. 

One option is to go back to SpanContext. If you have activateSpanContext, and deactivateSpanContext, this uncertainty disappears. Copying the spanContext off of a span now has the overhead of copying any underlying context, such as MDN, but this is perhaps the appropriate place for this. It feels semantically correct, and the overhead is perhaps minimal.

Example, with span context:

```java
  static class InstrumentedRunnable implements Runnable {
    private SpanContext spanContext;
    private Runnable wrappedRunnable;

    public InstrumentedRunnable(Runnable wrappedRunnable) {
      this.spanContext = GlobalTracer.get().activeSpan().context();
      this.wrappedRunnable = wrappedRunnable;
    }

    public InstrumentedRunnable(SpanContext spanContext, Runnable wrappedRunnable) {
      this.spanContext = spanContext;
      this.wrappedRunnable = wrappedRunnable;
    }

    @Override
    public void run() {
      tracer = GlobalTracer.get()
      tracer.activateContext(spanContext);
      try {
        wrappedRunnable.run();
      } finally {
	 tracer.deactivateContext(spanContext);
      }
    }
  }
```

@yurishkuro, I saw elsewhere that you were concerned about using spanContext instead of span for this, as it would be limiting. What do you think about this? It’s still a complete Span being managed via activeSpan, but it is only allowed to be set via a SpanContext. Is that that satisfactory?

The third option is to add **another** context-like concept to OpenTracing, just for activate/deactivate. I feel like that kind of… sucks.

**Comment by nicmunroe** [comment]
I would like to echo @mabn's concerns. This is a very real problem, at least in some use cases. I had to do this exact kind of state-saving and thread-hopping stuff with Wingtips (Zipkin/B3 compatible tracing library) in Riposte (Netty-based framework where async thread-hopping is the norm, not the exception).

I did a performance tuning pass recently and found that the MDC extraction/setup was a big performance hog if done excessively, and by tuning the code to only do it when absolutely necessary it more than doubled throughput in a hello-world type perf test. The MDC stuff went from being the biggest hotspot in CPU usage and garbage generation to almost invisible. 

Long story short - it can be tempting to try to shoehorn the state-saving/restoring and thread-hopping solution into a smaller API to prevent the complexity of the use case from spilling over into the API. It *feels* like there should be a solution where you don't have to expose those knobs, but I don't think you can properly address the use case without exposing it in the API. Saying "some of the state saving/restoring executions may be unnecessary but it's worth the tradeoff because it's likely to be only a few extra executions" is tempting but causes tracing to become extremely expensive in highly asynchronous situations.

There may be multiple solutions and it would be good to work towards minimizing confusion and accidental misuse of the API, but I don't think this use case can be sidestepped entirely.

**Comment by bensigelman** [comment]
@sjoerdtalsma:

> by the way, is this becoming too Java specific? I doubt there's an MDC concept in all languages

IMO it's great that we're talking about MDC, as it's both specific and problematic. My sense is that there are three categories of languages we're going to deal with as far as this issue is concerned:
- Things like PHP and Ruby(-on-Rails), where there's essentially only one request active at a time and very strongly established conventions about where to stash context information. They will be relatively easy.
- Things like Go, where there's only one way to pass context around (for better or worse :), so the question answers itself. This is why OpenTracing-Go is the only lib that already integrates with a context-propagation mechanism already. (I do think that integration could be better/cleaner, but that's for another day)
- Things like Java, where there are many ways to go about things and a cloudy sense of what's best. Of these languages (Java, Python, Node.js), Java seems like the hardest nut to crack... so getting something satisfactory in place would bode well for other languages.

**Comment by wu-sheng** [comment]
>One option is to go back to SpanContext. If you have activateSpanContext, and deactivateSpanContext, this uncertainty disappears. 

@tedsuo , back to which object should in charge of propagation. `SpanContext` is a liitle better than `Span` for inter-thread propagation, in understanding level. Still do not forget whether `SpanContext` or `Span`or anything else,  propagation through thread, we will face `the concurrency issue`. 

IMO, I mentioned before, why not we use the `carrier` concept (like inter-process propagation) to propagate? It is a existed concept, and everyone knows it will carry things for context propagation. `inter-process` and `inter-thread` should not be so different. More concepts mean more difficult to understand.

Even, we should use `SpanContext` in creating `Span`, or use `GlobalTracer`, that should another story.

**Comment by wu-sheng** [comment]
IMO. If you want to use the COPY of `span` or `SpanContext` to avoid `the concurrency issue` (like `MDC.getCopyOfContextMap()`), it will work, but not good enough. And it will bring more liabilities for `span` or `SpanContext`. These two concepts is already very clear.
- `Span` should be only for `start`, `finish`, `tag`, `log`
- `SpanContext` should be only for holding the context, but not propagate it. (cross-process propagation carrier exists)

The concept definitions are more important than API styles. Let's keep the class's duty precise.

 For users, they must understand  `the concurrency issue` , which clearly not necessary.

@bensigelman @sjoerdtalsma @tedsuo 

btw @pavolloffay I think you should participate in this. ☺

**Comment by JonathanMace** [comment]
@mabn FWIW this is what Baggage Buffers is for https://github.com/JonathanMace/tracingplane

It's a general-purpose request context object, but fundamentally supports `branch()` to duplicate contexts and `merge(a, b)` to combine two contexts into one.  It provides a bunch of different datatypes with built-in branch/merge semantics (eg, a set datatype with default merge of set union; true-wins flag, false-wins flag, counters, etc.)

I think it gives some of the other properties people discuss; its cross language, and you the set of concepts a person needs to understand/implement to be able to propagate baggage is very small.

I'll be talking about it at the workshop on Monday

**Comment by JonathanMace** [comment]
Here are my thoughts on what I _think_ the issue is, and I apologize in advance if I've misinterpreted the conversation

__________

I think the propagating contexts is so fundamentally distinct from managing spans that they should be considered separate components entirely.

Let's ignore spans for a moment and just talk about request contexts.

Let's treat a request context as an opaque object.  Who knows what it contains -- not us.  The basic goal we want to achieve is to propagate the request contexts along the execution path of requests -- across thread boundaries, with callbacks and continuations, maybe store them in thread-locals -- whatever.  Remember -- no spans!

I think doing this is straightforward.  The suggestions people have made earlier in this thread apply (eg, manager objects) apply perfectly well here.
The only APIs you need for a request context, in theory, is a way to `get` the current request context, and a way to `set` (or, `update`) the current request context.
Internally, the request context also needs semantics for `branch` (to copy to new threads, for example) and `join` (to merge multiple parent contexts, eg if threads join), but that is encapsulated by the request context implementation.

Now, suppose we've done this -- we have request context objects propagated everywhere we need.  Now we can add an API; lets say for now it's string key-value pairs (for ease of example)

Now the question is whether we can use this API to do spans?

So first, how do people use spans?  Well, it's straightforward -- you start them, do logs and annotations, then stop them.  I don't think it is unreasonable to ask users to manage spans explicitly in the code, if all you are doing is `startSpan`, `log`, `log`, ..., `endSpan`.

But, we also have to deal with:

1. how do we pass IDs between invocations of `startSpan` in order to capture causal relationships?
2. how do we pause a span in one place and resume it later on in another?

If we have a general-purpose request context, both of these become easy.

The span API will encapsulate the following logic in the single `startSpan` call:

1. check the request context for a `parentSpanId` field
2. set the span's parent ID if the `parentSpanId` is present
3. generate a new ID for the new span
4. set `parentSpanId` in the request context to the ID of the new span

Then, the ID will just propagate elsewhere, and the next time somebody calls `startSpan`, it will correctly capture the causal relationship.

As for pausing and resuming -- well, I believe this is also straightforward.

First, let's add `pauseSpan` and `resumeSpan` API calls.

The `pauseSpan` API internally does the following:

1. set a `currentSpanId` field to be the current span's ID

The `resumeSpan` API internally does the following:

1. create a new span object, but set its span ID to the value of `currentSpanId` (generate a new id or log an error if no `currentSpanId` exists)
2. clear `currentSpanId`

We also extend `startSpan` to do the following:

1. clear `currentSpanId`


------------

So that's what I think the high-level logic is.  In terms of discussing abstractions, does this seem like a good way to separate the concepts?

I suppose the remaining problem is:

1. Do you want additional state attached to Span objects?  When you pause a span in one place, do you want to be able to include a bunch of this hidden state, so that when you later resume it, you also restore the hidden state?

**Comment by bensigelman** [comment]
I spent a few hours on a plane flight experimenting with this. I have a bunch of code, but I don't feel good enough about the details to share it quite yet.

@JonathanMace I don't completely follow your pause/resume suggestion (esp how pausing *sets* a currentSpanId field (where?)). I am also troubled by the idea of creating a new Span (with its own timer) as part of "resumeSpan"... if we are pausing/resuming as, say, part of an async / event-based programming model, the Span start and finish times should represent something that's semantically important to the programmer. It's not uncommon in systems that do a bunch of RPCs for a single Span (say, representing a request handler) to be paused and resumed many times before it's finished. But maybe I'm just misunderstanding your proposal?

In any case, the crux of my investigations on the :airplane: revolved around the need to accommodate other non-tracing contextual data in a way that's still sufficiently performant. This is basically what @mabn was asking about.

I like the idea of something like this:

```
public interface ActiveSpanManager {
    // Basically a marker interface
    public interface Snapshot {
        Span span();
    }

    // Get the currently active Span (perhaps for this thread, etc)
    Span active();
    
    // Create a Snapshot encapsulating both the given Span and any state needed to activate/deactivate (see below)
    Snapshot snapshot(Span span);
    
    // Make the Snapshot and the Span it contains "active" per active().
    //
    // *Must* be paired with a subsequent call to deactivate().
    Span activate(Snapshot snapshot);
    
    // See activate() above.
    void deactivate(Snapshot snapshot);
}
```

(forgive any Java style problems)

In order to actually make this work without a bunch of wrappers that (IMO) will add boilerplate and confusion, it would also be helpful to add something like `Span.isFinished()`... otherwise it's difficult to know which Span to make `active()` when the currently-active one `deactivate()`s.

**Comment by tedsuo** [comment]
Thanks @bhs, this interface is definitely a nice refinement on the approach under discussion in this thread. 

Thoughts:
- I like that it is consolidated into a single API. By moving all the methods off of `Span` and `Tracer`, this allows most tracer implementations to take in an ActiveSpanManager at construction, and no wrapping is required for this basic relationship.  ActiveSpanManagers can be wrapped and combined, Tracers can be wrapped and combined, and these are handled separately.
- Tracer.getActiveSpanManager() should never be null, and be a NoopActiveSpanManager by default. Life will suck for framework authors otherwise.
- I like that the Snapshot carries the Span, and the Span does not carry the Snapshot. That was feeling very inverted to me.
- `activate` should not return `Span`, as the `Snapshot` now has the Span that is passed to `deactivate`.
- ActiveSpanManager implementations are for managing context information, such as `MDC`, but must also manage the stack of active spans. Multiple stacks of active spans doesn’t make sense, as they always must return the same active span. I’m concerned this makes combining context managers difficult. It should be one ActiveSpanManager that has multiple SnapshotManagers. Maybe that can happen behind the scenes of ActiveSpanManager.
- If this ActiveSpanManager interface works out, what does this say about the current SpanContext/Carrier abstraction? In-process prorogation and inter-process propagation feel like duals to me, I’m suspicious that their abstractions are shaped differently. Not for right now, but worth thinking about…

@JonathanMace, thanks for your comments. I've been very interesting in seeing an OT implementation on top of your Tracing Plan abstraction, so I'm happy to see https://github.com/JonathanMace/tracingplane-opentracing. In regards to your question about additional state being attached to the span: you are correct. It's a compatibility issue: OT needs to be compatible with a range of existing context management schemes, some of which (such as MDC) need to perform clone and merge operations before and after a context switch has occurred.

**Comment by pauldraper** [comment]
Pause/resume only matters if you want to profile, like "how much time was time spent in CPU", not counting time waiting on async. But I think that's outside what we want.

After using SpanManger, I realized I do not like stacks.

```java
ExecutorService executor = SpanPropagatingExecutors.newSingleThreadExecutor(DefaultSpanManager.getInstance);

void main() {
  executor.submit(new Runnable {
    public void run() {
      Thread.sleep(1);
      main();
    }
  })
}
```

This has a memory leak, since it tracks the (unbounded) async stack.

Am I missing a use case? Could the interface not be as simple as

```java
public interface SpanManger {
  void execute(Runnable f);
  void <T> execute(Callable<T> f);
}
// noticed downside: this doesn't work well with Java checked exceptions :/
```

**Comment by sjoerdtalsma** [comment]
@pauldraper I am curious about your simplification and trying to follow your thoughts.
How do you control which executor actually executes the `Runnable` and `Callable`? Is that decision implicitly embedded into the SpanManager?
If it doesn't execute in another thread, why manage the Span at all? Wouldn't a getCurrent() / setCurrent() be sufficient in that case?

**Comment by pauldraper** [comment]
No executor here. Perhaps I should have used `java.util.function.Supplier`. (Why you make things so confusing Java... </rant>).

Anyway,

```java
class ThreadLocalSpanManager {

  private void Span current = new ThreadLocal<Span> {
    protected Span initialValue() {
      return NoOpSpan.INSTANCE;
    }
  }

  Span getCurrent() {
    return current.get();
  };

  void execute(Span span, Runnable f) {
    Span former = current.get();
    current.set(span);
    try {
      f.run();
    } finally {
      current.set(former);
    }
  }

  void <T> execute(Span span, Callable<T> f) {
        Span former = current.get();
    current.set(span);
    try {
      return f.call();
    } finally {
      current.set(former);
    }
  }
}
```

Usage

```java
spanManager.execute(span, new Runnable() {
  public void run() {
    System.out.println(spanManager.getCurrent());
  }
});
```

The advantage to this API is simplicity which means (1) it's impossible to use incorrect (e.g. invoke this method for every that method) and (2) there are fewer caveats, like the memory leak I posted previously.

The only disadvantage would be if this API lacked some capability. Does it?

**Comment by bhs** [comment]
I have been pecking away at a POC implementation here: https://github.com/bhs/opentracing-java/pull/2/files

I deliberately did this off of the main opentracing-java repo because the code is so early. Please feel free to take a look. I'll paste the important bits in here, too.

## Code length

Don't be scared: the changes to `opentracing-api` are small. The rest of the code exists so that I could verify how stuff would get used. It would have been better to have really nice TDD for this, but I wanted to get basic feedback before doing anything clean.

## Tracer.buildSpan changes

This comment justifies all of the effort here:

```
   * <p>If there is an active Span according to the activeSpanManager(),
   * buildSpan will automatically have an asChildOf() reference to same.
   *
```

(BTW, I could imagine a new reference type to clarify that the association is implicit rather than explicit... but that's not as important as figuring out the SpanManager stuff for now.)

## io.opentracing.SpanManager

The most interesting part is probably the `SpanClosure` concept (not attached to the name!).

Also, @pauldraper, your approach is really appealing but I think (?) it breaks down in practice since there are plenty of situations where we won't have the luxury of wrapping everything in a `Runnable` or `Callable`... e.g., an interceptor framework where we get to run two separate blocks of code: one before `<something>` and one after it.

```
package io.opentracing;

/**
 * SpanManager allows an existing (possibly thread-local-aware) execution context provider to also manage the current
 * active OpenTracing span.
 */
public interface SpanManager {

    /**
     * A SpanClosure can be used *once* to make a Span active within a SpanManager, then deactivate it once the
     * "closure" (or period of Span activity) has finished.
     *
     * Most users do not directly interact with SpanClosure, activate(), or deactivate(), but rather use
     * SpanManager-aware Runnables/Callables/Executors. Those higher-level primitives need not be defined within the
     * OpenTracing core API.
     */
    interface SpanClosure {

        /**
         * Make the Span encapsulated by this SpanClosure active and return it.
         *
         * XXX: think harder about the returned Span being null.
         *
         * @see SpanManager#captureActive()
         * @return the newly-activated Span
         */
        Span activate();

        /**
         * End this active period for the Span previously returned by activate(). Finish the span iff finish=true.
         */
        void deactivate(boolean finishSpan);
    }

    /**
     * @return the currently active Span for this SpanManager, or null if no such Span could be found
     */
    Span active();

     /**
     * Capture any SpanManager-specific context (e.g., MDC context) along with the active Span (even if null) and
     * encapsulate it in a SpanClosure for activation in the future, perhaps in a different thread or on a different
     * executor.
     *
     * If the active Span is null, the implementation must still return a valid SpanClosure; when the closure activates,
     * it will clear any active Span.
     *
     * @see SpanManager.SpanClosure
     *
     * @return a SpanClosure that represents the active Span and any other SpanManager-specific context, even if the
     *     active Span is null.
     */
    SpanClosure captureActive();

    /**
     * Explicitly capture the given Span and any active state (e.g., MDC state) about the current execution context.
     *
     * @param span
     * @return a SpanClosure that represents the active Span and any other SpanManager-specific context, even if the
     *     active Span is null.
     */
    SpanClosure capture(Span span);

    /**
     * Tell the SpanManager that a particular Span has finished (and update any structures accordingly).
     */
    void onFinish(Span span);
}
```

## Active Span stacks

There is no explicit stack in the SpanManager (as there was in @sjoerdtalsma's contrib one)... the only stack is an implicit linked list in the `MDCSpanManager` demo's `SpanClosure` implementation.

## Other stuff

I added some comments to my own PR / POC... take a look there.

## In closing...

Please do take a look. Please don't judge my Java code or the level of polish here... this POC is not about naming/etc, but just about structure.

**Comment by wu-sheng** [comment]
What is the difference between `Snapshot` and `SpanClosure`? Seem two very similar concepts for me. You want to use the `SpanClosure` to build the context between threads, since the `ThreadLocal` did not work.

**Comment by objectiser** [comment]
@bhs If transferring a `Span` from one thread to another, by creating a `SpanClosure` which is activated in the destination thread, how is the `Span` disassociated from the first thread (in situations where this is desirable)?

**Comment by bhs** [comment]
@wu-sheng 

> What is the difference between Snapshot and SpanClosure?

Definitely similar in spirit... The SpanClosure actually has an interface to implement, though, since it makes it clear that a SpanClosure/Snapshot may only be activated/deactivated once.

**Comment by bhs** [comment]
@objectiser 

> If transferring a Span from one thread to another, by creating a SpanClosure which is activated in the destination thread, how is the Span disassociated from the first thread (in situations where this is desirable)?

Yes, that's a good point. I had addressed that in a previous version of this (that I didn't send out) where everything was reference-counted and `finish()` was implicit (when the refcount got to zero).

I've added [a new commit](https://github.com/bhs/opentracing-java/pull/2/commits/555f5ea9ee96972c9324d80dabc20adacfb4fbd6) that includes a new method, `SpanBuilder.startAndActivate()`, that returns a `SpanClosure`. I also [made `SpanClosure` autocloseable](https://github.com/bhs/opentracing-java/pull/2/commits/396abd2e085a5fcd1871146d37c38c2c5d91dec8) just to see how that felt syntactically.

The async code looks like this:

```
        // Create a parent SpanClosure for all of the async activity.
        try (SpanManager.SpanClosure parentSpanClosure = tracer.buildSpan("parent").startAndActivate();) {

            // Create 10 async children.
            for (int i = 0; i < 10; i++) {
                final int j = i;
                futures.add(otExecutor.submit(new Runnable() {
                    @Override
                    public void run() {
                        // START child body

                        try (SpanManager.SpanClosure childSpanClosure =
                                     tracer.buildSpan("child_" + j).startAndActivate();) {
                            Thread.currentThread().sleep(1000);
                            childSpanClosure.span().log("awoke");
                            Runnable r = new Runnable() {
                                @Override
                                public void run() {
                                    Span active = tracer.activeSpanManager().active();
                                    active.log("awoke again");
                                    // Create a grandchild for each child.
                                    Span grandchild = tracer.buildSpan("grandchild_" + j).start();
                                    grandchild.finish();
                                    active.finish();
                                }
                            };
                            subfutures.add(otExecutor.submit(r));
                        } catch (Exception e) { }

                        // END child body
                    }
                }));
            }
        } catch (Exception e) { }

        try {
            for (Future<?> f : futures) {
                f.get();
            }
            for (Future<?> f : subfutures) {
                f.get();
            }
        } catch (Exception e) { }
```

It's worth noting that existing synchronous start()/finish() code can work just as it does today.

**Comment by sjoerdtalsma** [comment]
One question; are we 'okay' now with automatically/implicitly _becoming child of the active span_ ?
I remember rather strong opinions on either side of that decision when discussing the java-spanmanager PR.

**Comment by objectiser** [comment]
@bhs There may still be an issue if the parent scope is being managed by two separate event handlers, if the event handler creating the `SpanClosure` cannot store it in some specific context that is available to the end event handler.

**Comment by bhs** [comment]
@sjoerdtalsma 

> One question; are we 'okay' now with automatically/implicitly becoming child of the active span ?
I remember rather strong opinions on either side of that decision when discussing the java-spanmanager PR.

I'm definitely ok with that. We did so in Dapper and – while it has some drawbacks – it's a way better default than not-becoming the child :)

Two thoughts, though:
1. *If.f.* the programmer explicitly provides references, those would replace (not supplement) any automatic references, and
2. We can introduce a new Reference type, maybe `INFERRED_CHILD_OF` or similar; this will allow tracing systems / UIs / etc to distinguish these relationships


@objectiser 

> There may still be an issue if the parent scope is being managed by two separate event handlers, if the event handler creating the SpanClosure cannot store it in some specific context that is available to the end event handler.

IMO, if different executors (or whatever) start and finish a Span, they would not share a single SpanClosure activation/deactivation. Either there would be zero or two of those pairs. Note that in my proposal it's still possible to start a Span without meddling with SpanManager at all.

**Comment by bhs** [comment]
A little later than I'd hoped, but here is a sincere attempt to address this issue for OT-Java: https://github.com/opentracing/opentracing-java/pull/111

Please add comments there... I'm making special note of it because more people are watching this issue.

**Comment by beberlei** [comment]
The java implementation scares me, since this kind of complexity is absolutely not needed in a request scoped language such as PHP. Would it make more sense to agree that each language that does not have explicit in process context propagation such as Go, should provide a solution that is idiomatic for this language?

**Comment by felixfbecker** [comment]
Since I hear this a lot, not every PHP project is request scoped. PHP can be used just like any other language to build long-running service processes, that e.g. can serve RPCs calls over TCP, even with an event loop like in NodeJS. Take as an example https://github.com/felixfbecker/php-language-server

**Comment by yurishkuro** [comment]
@beberlei yes, the propagation mechanism is specific to each language, although it is preferable to keep similar concepts as much as possible, such as "active span".

@felixfbecker if you look at a proposed mechanism for Python, it is actually a pluggable approach depending on which async framework in Python is used, because they require different ways of propagating the context via event loops. I assume the same will be true in PHP. For request-scoped PHP the implementation of active span source could be just a global variable.

**Comment by felixfbecker** [comment]
For request-scoped PHP that is not using async functions with callbacks (like async curl or async DB queries) and no event loops, it can be a global variable.

A runtime also doesn't necessarily need to have a concept of "active span" imo. In Node's event loop continuation-local-storage is super slow to propagate an active span, and it is so much easier to e.g. just pass the span around as the last parameter to functions.

**Comment by dkuebric** [comment]
@felixfbecker without an active span getter, it will be more cumbersome for users to combine their own instrumentation points with those built in to standard libraries (eg. web frameworks, db clients, etc), and to compose those with each other.

**Comment by tedsuo** [comment]
@yurishkuro has the right nuance here. In-process propagation is language dependent, but for languages that share paradigms we should try to avoid unnecessary inconsistency between API implementations. For example, I would not want the ruby implementation to stray too far from the python one.

@felixfbecker I think Javascript is a special case, where you do really want an ActiveSpan concept but the currently available mechanisms are insufficient/unofficial enough that it would be concerning to bake support for them into the official API. But if JS _were_ to sort out continuation local storage in an effective manner, we would immediately want it for the reasons @dkuebric pointed out.

PHP concerns me because I'm ignorant, and don't understand the various possible execution contexts and how they are exposed to the PHP runtime, and to what degree the tracing you need to do is actually mixed up with foreign function calls to C code.
